src/NUnitEngine/nunit.engine.tests/Transport/Tcp/TcpServerTests.cs

```diff
@@ -28,42 +30,58 @@ public void StopServer()
-        public void SingleClientConnection()
+        [TestCase(1)]
+        [TestCase(3)]
+        [TestCase(20)]
+        public void ClientConnectionTest(int numClients)
         {
-            using (TcpClient client = new TcpClient())
+            var clients = new TcpClient[numClients];
+            var workers = new BackgroundWorker[numClients];
+
+            var allTasksComplete = new ManualResetEvent(false);
+            int pendingTasks = numClients;
+
+            for (int i = 0; i < numClients; i++)
             {
-                client.Connect(_server.EndPoint);
-                client.Client.Send(new Guid().ToByteArray());
+                var client = new TcpClient();
+                clients[i] = client;
+
+                using (var worker = new BackgroundWorker())
+                {
+                    worker.DoWork += (s, e) =>
+                    {
+                        client.Connect(_server.EndPoint);
+                        client.Client.Send(new Guid().ToByteArray());
+                    };
 
-                Thread.Sleep(1); // Allow the connection event to run
-                Assert.That(_serverConnections.Count, Is.EqualTo(1), "Should have received 1 connection event");
-                Assert.That(_serverConnections[0].Connected, "Server is not connected to client");
+                    worker.RunWorkerCompleted += (s, e) =>
+                    {
+                        if (Interlocked.Decrement(ref pendingTasks) == 0)
+                            allTasksComplete.Set();
+                    };
 
-                Assert.That(client.Connected, Is.True, "Client is not connected to server");
+                    workers[i] = worker;
+                }
             }
-        }
 
-        [Test]
-        public void MultipleClientConnections()
-        {
-            TcpClient[] clients = new[] { new TcpClient(), new TcpClient(), new TcpClient() };
-            int num = clients.Length;
+            foreach (var worker in workers)
+                worker.RunWorkerAsync();
 
-            foreach (var client in clients)
-            {
-                client.Connect(_server.EndPoint);
-                client.Client.Send(new Guid().ToByteArray());
-            }
+            allTasksComplete.WaitOne();
 
-            Thread.Sleep(1); // Allow the connection events to run
-            Assert.That(_serverConnections.Count, Is.EqualTo(num), $"Should have received {num} connection events");
+            Assert.That(_serverConnections.Count, Is.EqualTo(numClients), $"Should have received {numClients} connection events");
 
-            for (int i = 0; i < num; i++)
+            for (int i = 0; i < numClients; i++)
             {
                 Assert.That(_serverConnections[i].Connected, $"Server is not connected to client {i + 1}");
                 Assert.That(clients[i].Connected, Is.True, $"Client {i + 1} is not connected to server");
             }
         }
+
+        private void RunTask(TcpClient client)
```

Comment:
Was this supposed to be used in the DoWork above?

Reply:
Left over from the code I replace.